src/compiler/checker.ts

```diff
@@ -24121,6 +24111,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const templateType = getTemplateTypeFromMappedType(target);
         const inference = createInferenceInfo(typeParameter);
         inferTypes([inference], sourceType, templateType);
+        const sourceValueDeclaration = sourceType.symbol?.valueDeclaration;
+        if (sourceValueDeclaration) {
+            const intraExpressionInferenceSites = getInferenceContext(sourceValueDeclaration)?.intraExpressionInferenceSites?.filter(site => isNodeDescendantOf(site.node, sourceValueDeclaration));
+            if (intraExpressionInferenceSites?.length) {
+                const templateType = (getApparentTypeOfContextualType(sourceValueDeclaration.parent.parent as Expression, ContextFlags.NoConstraints) as MappedType).templateType;
+                if (templateType) {
+                    Debug.assert(isExpressionNode(sourceValueDeclaration));

```

Comment:
I'm not fond of the cast to `Expression` in the line below this. That's why I've added this `Debug.assert` here. If I reason about this correctly... with the correct checks before this line and all it should be 100% guaranteed that `sourceValueDeclaration` is an expression. Otherwise, we wouldn't find any "relevant" intra expression inference sites.

I'm not sure what's the codebase policy around stuff like that so if you have any suggestions on how this should be done here I'm all ears

Reply:
I think you're looking for Debug.assertNode?

Reply:
Thanks for the type, I pushed out a change to use it + I changed `isExpressionNode` to be an `asserts` function.